### ISSUE TITLE
Improve mouse wheel normalization code to adjust scrolling sensitivity across browsers

### ISSUE THREAD

**Issue by vvandens** [qooxdoo/qooxdoo#9303]
This issue has been exposed on [Stackoverflow](http://stackoverflow.com/questions/42872306/adjusting-qooxdoo-scrolling-speed-sensitivity) with the premise of a solution by @level420 :

I'm wondering if it is possible to adjust the scrolling sensitivity in qooxdoo applications using qx.ui.container.Scroll or any internally scrollable components, e.g. qx.ui.table.Table.

What I'm experiencing on my MacBook Pro running Sierra 10.12.3 is a blazing fast scrolling speed using the trackpad, and it only appears to be the case in qooxdoo-based applications which makes them very difficult to use. Tried different browsers with comparable results (Chrome, Safari, ...).

As an example, open the [API browser](http://www.qooxdoo.org/5.0.2/api/index.html) and scroll the browsing API tree on the left.

**Comment by level420** [comment]
@vvandens thank you for creating this issue.

The following thread in the apple forum gives some additional information https://discussions.apple.com/thread/7679256?start=0&tstart=0

**Comment by level420** [comment]
@vvandens could you please verify if the following hint solves this?
https://discussions.apple.com/message/30700725#30700725

Just to be sure it's not the problem described in that thread.
Thank you.

**Comment by vvandens** [comment]
@level420 Unfortunately, it doesn't solve anything. I'll take some time to play with the code and see how it goes.

**Comment by vvandens** [comment]
I've taken a quick look to the code. My (little) understanding of the problem is that the way `WheelEvent`s are handled cannot be directly applied to the MacBook Pro trackpad events. Don't know however if the same problem arise on other hardware/os.

After some debugging, I found that the constant `qx.event.handler.GestureCore.ROLL_FACTOR`, that is hardcoded to `18`, is used to determine the number of pixels to actually scroll depending on the number of mouse wheel ticks (i.e. 18 pixels / tick ?). However, the mbp trackpad fires a lot of such `WheelEvent`s; much more than a mouse wheel would actually do which explains the very high perceived sensitivity.

So the next step is to find a way to distinguish between actual mouse wheel events and trackpad generated ones in order to be able to apply a different `ROLL_FACTOR`. My testings show that the event [`deltaMode`](https://developer.mozilla.org/en-US/docs/Web/API/WheelEvent/deltaMode) is set to DOM_DELTA_PIXEL in case of a trackpad event and DOM_DELTA_LINE in case of a mouse wheel event. So my first idea would be to split the ROLL_FACTOR constant to LINE_ROLL_FACTOR (`18`) and `PIXEL_ROLL_FACTOR` (for which `2` seems to be a sensible value).

Then we need to test other hardware/os combination in order to determine if this rule is fully portable or should be restricted to the Mac OS. Unfortunately, I'm short on this one.

What do you think ?

**Comment by johnspackman** [comment]
@vvandens good detective work 👍 

It looks like a common problem; I've found this SO post [1] which has a way to differentiate, not ideal but it might help

[1] http://stackoverflow.com/a/31191250

**Comment by vvandens** [comment]
Thanks @johnspackman . We could end up with such a solution as described in SO, but I would prefer to stick with [`deltaMode`](https://developer.mozilla.org/en-US/docs/Web/API/WheelEvent/deltaMode) switching if it is enough.
However, I don't know how to check that all real mouse wheels will generate an event with deltaMode=DOM_DELTA_LINE and all trackpads with deltaMode=DOM_DELTA_PIXEL, although this is what I experienced.

**Comment by vvandens** [comment]
Forget it... It just doesn't work. My eyes lied. Both wheel and trackpad generate events with deltaMode=DOM_DELTA_PIXEL, at least on Mac OS. So I think we have to dig the SO solution. 

However, I must admit that I'm a little lost as what is the best place to put this type of code into the qooxdoo framework... Maybe simply in the `qx.event.handler.GestureCore._fireRoll` method ?

**Comment by level420** [comment]
@vvandens I think it would be best if the changes happen here : https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/util/Wheel.js

`qx.event.handler.GestureCore._fireRoll` calls in first place `qx.util.Wheel.getDelta(domEvent, "x")`

But it may be that we have to change the static `qx.event.handler.GestureCore.ROLL_FACTOR` to something dynamic.

**Comment by level420** [comment]
my favorite would be `qx.util.Wheel.__recalculateMultiplicator`.

If you need distinguish between the operating system and the os version, you'll get it via `qx.core.Environment.get("os.name")` and ``qx.core.Environment.get("os.version")` which deliver `osx` and `10.12`. This works on safari, chrome and firefox.

**Comment by vvandens** [comment]
@level420 OK. However, the SO code seems to be a one-time detection code, i.e. once it has determined that the device is a trackpad, it doesn't change anymore as shown in this [JSFiddle](https://jsfiddle.net/ucLe3hLa/1/).
Do we want this behavior or do we want a more dynamic one, where the user can switch between the mouse and the trackpad and the scroll sensitivity adapts dynamically ?

**Comment by vvandens** [comment]
I've come up with the following [pull request](https://github.com/qooxdoo/qooxdoo/pull/9304). Unfortunately, since the detection must happen before the event is actually fired, I had to implement it in the `qx.event.handler.GestureCore._fireRoll` method. The idea is to accumulate the count of events in order to detect if they come from a trackpad before actually delivering them to the application. Some remarks :

- The different constants might need some adjustments.
- I've not included any condition on browser / hardware / os / version , since I believe that this can be useful on different combinations we cannot be aware of. However, feel free to amend.

**Comment by vvandens** [comment]
I've just submitted a [new PR](https://github.com/qooxdoo/qooxdoo/pull/9306) to solve the issue. Waiting for your tests and comments.

**Comment by level420** [comment]
Fixed via #9306
